File: src/shesasyncer/alignment/sequence.py
from difflib import SequenceMatcher

from ..lyrics.normalize import normalize_lyric


def similarity(a: str, b: str) -> float:
    return SequenceMatcher(None, normalize_lyric(a), normalize_lyric(b)).ratio()


def monotonic_match(lyrics: list[str], segments: list[dict], *, skip_penalty: float = 0.18) -> list[tuple[int, int, float]]:
    """Globally align lyric lines to timed segments while preserving order.

    Unlike greedy matching, repeated choruses and locally similar lines can be
    resolved using surrounding context. The result contains only matched pairs.
    """
    n, m = len(lyrics), len(segments)
    if not n or not m:
        return []
    dp = [[0.0] * (m + 1) for _ in range(n + 1)]
    move = [[None] * (m + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            s = similarity(lyrics[i - 1], str(segments[j - 1].get("text", "")))
            diag = dp[i - 1][j - 1] + s
            up = dp[i - 1][j] - skip_penalty
            left = dp[i][j - 1] - skip_penalty
            best = max((diag, "diag"), (up, "up"), (left, "left"))
            dp[i][j], move[i][j] = best
    out = []
    i, j = n, m
    while i and j:
        if move[i][j] == "diag":
            s = similarity(lyrics[i - 1], str(segments[j - 1].get("text", "")))
            if s > 0:
                out.append((i - 1, j - 1, s))
            i -= 1; j -= 1
        elif move[i][j] == "up":
            i -= 1
        else:
            j -= 1
    return list(reversed(out))

File: src/shesasyncer/alignment/sequence.js
import { normalizeLyric } from '../lyrics/normalize.js';

export function similarity(a, b) {
  const x = normalizeLyric(a), y = normalizeLyric(b);
  if (x === y) return x ? 1 : 0;
  if (!x || !y) return 0;
  const prev = Array(y.length + 1).fill(0).map((_, i) => i);
  for (let i = 1; i <= x.length; i++) {
    let left = i;
    for (let j = 1; j <= y.length; j++) {
      const old = prev[j];
      const cost = x[i - 1] === y[j - 1] ? 0 : 1;
      prev[j] = Math.min(prev[j] + 1, left + 1, prev[j - 1] + cost);
      left = prev[j];
    }
  }
  const distance = prev[y.length];
  return 1 - distance / Math.max(x.length, y.length);
}

export function monotonicMatch(lyrics, segments, { skipPenalty = 0.18 } = {}) {
  const n = lyrics.length, m = segments.length;
  if (!n || !m) return [];
  const dp = Array.from({ length: n + 1 }, () => Array(m + 1).fill(0));
  const move = Array.from({ length: n + 1 }, () => Array(m + 1).fill(null));
  for (let i = 1; i <= n; i++) for (let j = 1; j <= m; j++) {
    const s = similarity(lyrics[i - 1], segments[j - 1]?.text ?? '');
    const choices = [[dp[i - 1][j - 1] + s, 'diag'], [dp[i - 1][j] - skipPenalty, 'up'], [dp[i][j - 1] - skipPenalty, 'left']];
    choices.sort((a, b) => b[0] - a[0]);
    [dp[i][j], move[i][j]] = choices[0];
  }
  const out = []; let i = n, j = m;
  while (i && j) {
    if (move[i][j] === 'diag') { const s = similarity(lyrics[i - 1], segments[j - 1]?.text ?? ''); if (s > 0) out.push([i - 1, j - 1, s]); i--; j--; }
    else if (move[i][j] === 'up') i--; else j--;
  }
  return out.reverse();
}
